Validation

nassersaazi opened the pull request:
Closes #10 

- Ensure a transcript is not uploaded more than once
- Ensure POST requests for transcripts are coming from authorized source
- Update database seed data to match updated changes

nassersaazi committed:
Make adjustments to Transcript model

- Change default status from 'not queued' to 'queued'
- Don't allow null values for 'originalContent' field

Add hash generation logic

Add validation logic for transcripts

nassersaazi committed:
Add functionality for auth token

nassersaazi committed:
Add validation logic for transcripts

- Ensure a transcript is not uploaded more than once
- Ensure POST requests for transcripts are coming from authorised source
- Update database seed data to match updated changes

Remove redundant function

Emmanuel-Develops commented on api/app/controllers/transcript.controller.js:
```diff
@@ -1,23 +1,61 @@
-    // We have to add title because for some reason having just content makes an update to an existing record insted of inserting a new one
     originalContent: req.body.originalContent,
-    content: req.body.content
+    content:req.body.originalContent,
```
content and originalContent both refer to `req.body.originalContent` why is that?

nassersaazi replied:
nice catch. going to fix that...

Emmanuel-Develops commented on api/app/controllers/transcript.controller.js:
```diff
@@ -1,23 +1,61 @@
+
+    const oc = req.body.originalContent
+    const hashParams = oc.title + oc.media + oc.date
+    const transcriptHash = crypto.createHash('sha256').update(hashParams).digest('base64');
```
It could be a security risk using a static token, also since we are including hashParams (oc.title+oc.media+oc.date) the token will be different for each payload. Would the static TOKEN work?
We can have a shared secret then hash it with that secret and the payload parameters then compare the hashes.

```js
const key = crypto.createHash('sha256').update(secret).digest();
const hashParams = oc.title + oc.media + oc.date
const transcriptHash = crypto.createHash('sha256', key).update(hashParams).digest('base64');
```

nassersaazi replied:
so i think we are confusing two different things here. the `transcriptHash` and the `authToken` .For the transcript hash, it should be unique for every transcript to be double posting. For the auth token .I initially had a naive implementation of a static token ,but given what we've discussed on the call ,am going to implement a hash of a shared secret and have it passed through the header. Incidentally, that would also fix #1

nassersaazi replied:
isn't there a possibility that the media source in the JSON payload could be null? in that case do we just dismiss the transcript?

adamjonas replied:
if that's the case, I'd do a match on title + first 5 words of the body. I think a hash is overkill given that whitespace and casing with change all the hashes and I'd prefer to avoid the dependency.

adamjonas commented on api/app/config/auth.config.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  TOKEN: "936a185caaa266bb9cbe981e9e05cb78cd732b0b3280eb944412bb6f8f8f07af"
```
this token needs to be a secret

nassersaazi committed:
Enforce stricter checks on posted data

- Validate transcript hash on media source
- Pass auth secret as environment variable
- Fix issue with duplicate originalContent request parameter

nassersaazi committed:
Remove auth token logic

- Alternative logic for backend auth was discussed in discord

nassersaazi committed:
Use unique string instead of hash to prevent duplicate transcripts
